jplag.endToEndTesting/src/test/java/de/jplag/end_to_end_testing/JavaEndToEndTest.java

```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * Inserting comments or empty lines (normalization level) -> id = 0 Changing variable names or function names
+     * (normalization level) -> id = 1 Inserting comments or empty lines (normalization level) -> id = 2
+     * @param testClassNames Plagiarized classes names in the resource directorie which are needed for the test
+     * @param testId name of the testId to load and identify the stored results
+     * @throws IOException is thrown in case of problems with copying the plagiarism classes
+     * @throws ExitException in case the plagiarism detection with JPlag is preemptively terminated would be of the test.
+     */
+    @ParameterizedTest
+    @MethodSource("normalizationLevelTestArguments")
+    void normalizationLevelTest(String[] testClassNames, int testId) throws IOException, ExitException {
+
+        TestCaseModel testCaseModel = jplagTestSuiteHelper.createNewTestCase(testClassNames);
+
+        JPlagResult jplagResult = new JPlag(testCaseModel.getJPlagOptionsFromCurrentModel()).run();
+
+        var jsonModel = testCaseModel.getCurrentJsonModel();
+
+        for (JPlagComparison jPlagComparison : jplagResult.getAllComparisons()) {
+            assertEquals(jsonModel.getResultModelById(testId).getResultSimilarity(), jPlagComparison.similarity(),
+                    "The JPlag results [similarity] do not match the stored values!");
+        }
+    }
+
+    /**
+     * Insertion of unnecessary or changed code lines (token generation) -> id = 0 Changing the program flow (token
+     * generation) (statements and functions must be independent from each other) -> id = 1 Variable declaration at the
+     * beginning of the program (Detecting Source Code Plagiarism [...]) -> id = 2
+     * @param testClassNames Plagiarized classes names in the resource directorie which are needed for the test
+     * @param testId name of the testId to load and identify the stored results
+     * @throws IOException is thrown in case of problems with copying the plagiarism classes
+     * @throws ExitException in case the plagiarism detection with JPlag is preemptively terminated would be of the test.
+     */
+    @ParameterizedTest
+    @MethodSource("tokenGenerationLevelTestArguments")
+    void tokenGenerationLevelTest(String[] testClassNames, int testId) throws IOException, ExitException {
+
+        TestCaseModel testCaseModel = jplagTestSuiteHelper.createNewTestCase(testClassNames);
+
+        JPlagResult jplagResult = new JPlag(testCaseModel.getJPlagOptionsFromCurrentModel()).run();
+
+        var jsonModel = testCaseModel.getCurrentJsonModel();
+
+        for (JPlagComparison jPlagComparison : jplagResult.getAllComparisons()) {
+            assertEquals(jsonModel.getResultModelById(testId).getResultSimilarity(), jPlagComparison.similarity(),
+                    "The JPlag results [similarity] do not match the stored values!");
+        }
+    }
```

Comment:
The methods are equal. Simply create a private helper method that contains the code. Then call the method in each of the test methods

Reply:
Chagned to:
### private helper method
``` java 
    /**
     * This method creates the necessary results as well as models for a test run and summarizes them for a comparison.
     * @param testClassNames Plagiarized classes names in the resource directorie which are needed for the test
     * @return Results and necessary comparison values for the tests in the form of the TestResultModel
     * @throws IOException s thrown in case of problems with copying the plagiarism classes
     * @throws ExitException in case the plagiarism detection with JPlag is preemptively terminated would be of the test.
     */
    private TestResultModel runCurrentTestCase(String[] testClassNames) throws IOException, ExitException {
        String functionName = StackWalker.getInstance().walk(stream -> stream.skip(1).findFirst().get()).getMethodName();
        TestCaseModel testCaseModel = jplagTestSuiteHelper.createNewTestCase(testClassNames, functionName);
        JPlagResult jplagResult = new JPlag(testCaseModel.getJPlagOptionsFromCurrentModel()).run();
        var jsonModel = testCaseModel.getCurrentJsonModel();

        return new TestResultModel(jplagResult, jsonModel);
    }
```

### The associated test methods

``` java
/**
     * Inserting comments or empty lines (normalization level) -> id = 0 Changing variable names or function names
     * (normalization level) -> id = 1 Inserting comments or empty lines (normalization level) -> id = 2
     * @param testClassNames Plagiarized classes names in the resource directorie which are needed for the test
     * @param testId name of the testId to load and identify the stored results
     * @throws IOException is thrown in case of problems with copying the plagiarism classes
     * @throws ExitException in case the plagiarism detection with JPlag is preemptively terminated would be of the test.
     */
    @ParameterizedTest
    @MethodSource("normalizationLevelTestArguments")
    void normalizationLevelTest(String[] testClassNames, int testId) throws IOException, ExitException {

        TestResultModel testResultModel = runCurrentTestCase(testClassNames);

        for (JPlagComparison jPlagComparison : testResultModel.getJPlagResult().getAllComparisons()) {
            assertEquals(testResultModel.getCurrentJsonModel().getResultModelById(testId).getResultSimilarity(), jPlagComparison.similarity(),
                    "The JPlag results [similarity] do not match the stored values!");
        }
    }
```

Reply:
Isn't the latter 

```java 
TestResultModel testResultModel = runCurrentTestCase(testClassNames);
for (JPlagComparison jPlagComparison : testResultModel.getJPlagResult().getAllComparisons()) {
      assertEquals(testResultModel.getCurrentJsonModel().getResultModelById(testId).getResultSimilarity(), jPlagComparison.similarity(), "The JPlag results [similarity] do not match the stored values!");
```

the same for both tests as well?

Reply:
Yes the comparisons are identical. I could summarize both test methods also together however in the overview one would not be able to hold these then correctly apart. I could indicate into the methods parameter still another "use" what can replace then the method name. 

### from
``` java 
    /**
     * Test cases created for the normalization level
     * @return the classes to be tested with the corresponding identifier
     */
    private static Stream<Arguments> normalizationLevelTestArguments() {
        return Stream.of(Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo1.java"}, 0),
                Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo2.java"}, 1),
                Arguments.of((Object) new String[] {"SortAlgo1.java", "SortAlgo2.java"}, 2));
    }
```

### to
``` java
    
    /**
     * Test cases created for the normalization level
     * @return the classes to be tested with the corresponding identifier
     */
    private static Stream<Arguments> TestArguments() {
        return Stream.of(Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo1.java"}, 0, "NormalizationLevel"),
                Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo2.java"}, 1,  "NormalizationLevel"),
                Arguments.of((Object) new String[] {"SortAlgo1.java", "SortAlgo2.java"}, 2 ,  "NormalizationLevel"),
                Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo3.java"}, 0, "TokengenerationLevel"),
                Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo4.java"}, 1, "TokengenerationLevel"),
                Arguments.of((Object) new String[] {"SortAlgo.java", "SortAlgo4d1.java"}, 2, "TokengenerationLevel"));
    }
```

Reply:
I was happy with the two methods for the test arguments and the separation of the two groups. I would only suggest to create some code like this:

```java
@ParameterizedTest
@MethodSource("normalizationLevelTestArguments")
void normalizationLevelTest(String[] testClassNames, int testId) throws IOException, ExitException {
      runJPlagTestSuite(testClassNames, testId);
}

private void runJPlagTestSuite(String[] testClassNames, int testId) throws IOException, ExitException {
        TestCaseModel testCaseModel = jplagTestSuiteHelper.createNewTestCase(testClassNames);
        JPlagResult jplagResult = new JPlag(testCaseModel.getJPlagOptionsFromCurrentModel()).run();
        var jsonModel = testCaseModel.getCurrentJsonModel();
        for (JPlagComparison jPlagComparison : jplagResult.getAllComparisons()) {
            assertEquals(jsonModel.getResultModelById(testId).getResultSimilarity(), jPlagComparison.similarity(), "The JPlag results [similarity] do not match the stored values!");
        }
    }

```

Reply:
I have included the suggested changes in the code:

``` java

    @ParameterizedTest
    @MethodSource("tokenGenerationLevelTestArguments")
    void tokenGenerationLevelTest(String[] testClassNames, int testIdentifier) throws IOException, ExitException {
        runJPlagTestSuite(testClassNames, testIdentifier);
    }

    private void runJPlagTestSuite(String[] testClassNames, int testIdentifier) throws IOException, ExitException {
        String functionName = StackWalker.getInstance().walk(stream -> stream.skip(1).findFirst().get()).getMethodName();
        TestCaseModel testCaseModel = jplagTestSuiteHelper.createNewTestCase(testClassNames, functionName);
        JPlagResult jplagResult = new JPlag(testCaseModel.getJPlagOptionsFromCurrentModel()).run();
        var jsonModel = testCaseModel.getCurrentJsonModel();

        for (JPlagComparison jPlagComparison : jplagResult.getAllComparisons()) {
            assertEquals(jsonModel.getResultModelById(testIdentifier).getResultSimilarity(), jPlagComparison.similarity(),
                    "The JPlag results [similarity] do not match the stored values!");
        }
    }
```